benchmarks/htap/lib/controller.py

```diff
@@ -0,0 +1,167 @@
+        worker_id, dsn, scale_factor, dry_run, monitoring_interval, reporting_queue, date_value
+):
+    with DBConn(dsn) as conn:
+        start_date = datetime.fromtimestamp(date_value.value)
+        tpcc_worker = Transactions(worker_id, scale_factor, start_date, conn)
+        reporting_interval = timedelta(seconds=monitoring_interval) / 2
+        last_reporting_time = None
+        while True:
+            tpcc_worker.next_transaction(dry_run)
+            now = datetime.now()
+            if last_reporting_time is None or (now - last_reporting_time) >= reporting_interval:
+                reporting_queue.put_nowait(('tpcc', tpcc_worker.stats(worker_id)))
+                last_reporting_time = now
+
+
+def olap_worker(worker_id, dsn, olap_timeout_seconds, dry_run, reporting_queue, date_value):
+    with DBConn(dsn, statement_timeout=olap_timeout_seconds * 1000) as conn:
+        queries = Queries(worker_id, conn)
+        while True:
+            max_date = datetime.fromtimestamp(date_value.value)
+            queries.run_next_query(max_date, reporting_queue, dry_run)
+
+
+class HTAPController:
+    def __init__(self, args):
+        self.dsn = args.dsn
+        self.scale_factor = self._query_num_warehouses()
+        self.dry_run = args.dry_run
+        self.num_oltp_workers = args.oltp_workers
+        self.num_olap_workers = args.olap_workers
+        self.benchmark_duration = args.duration
+        self.olap_query_timeout = args.olap_timeout
+        self.monitoring_interval = args.monitoring_interval
+        self.collect_stats = args.stats_dsn is not None
+        self.stats_dsn = args.stats_dsn
+
+        self.stats = Stats(self.num_oltp_workers, self.num_olap_workers)
+        self.monitor = Monitor(
+                self.stats, self.num_oltp_workers, self.num_olap_workers, self.scale_factor
+        )
+
+        print(f'Detected scale factor: {self.scale_factor}')
+
+    def _sql_error(self, msg):
+        import sys
+        print(f'ERROR: {msg} Did you run `prepare_benchmark`?')
+        sys.exit(-1)
+
+    def _query_latest_delivery_date(self):
+        with DBConn(self.dsn) as conn:
+            try:
+                conn.cursor.execute('SELECT max(ol_delivery_d) FROM order_line')
+                return conn.cursor.fetchone()[0]
+            except psycopg2.ProgrammingError:
+                self._sql_error('Could not query the latest delivery date.')
+
+    def _query_num_warehouses(self):
+        with DBConn(self.dsn) as conn:
+            try:
+                conn.cursor.execute('SELECT count(distinct(w_id)) from warehouse')
+                return conn.cursor.fetchone()[0]
+            except psycopg2.ProgrammingError:
+                self._sql_error('Could not query number of warehouses.')
+
+    def _launch_workers(self, pool, queue, date_value):
+        def worker_error(err):
+            import sys
+            print(f'Exception in worker: {err}')
+            # FIXME: not all exceptions have __traceback__ set
```

Comment:
TODO

Reply:
still TODO for this PR?

Reply:
No, I couldn't find again which exception it was that had no __traceback__ property. But I decided to leave it in in case somebody comes across that problem again.